Sources/Networking/Core/DownloadAPIManager.swift

```diff
@@ -0,0 +1,220 @@
+
+             /// reset retry count
+             await retryCounter.reset(for: endpointRequest.id)
+
+             /// create download AsyncStream
+             return (downloadTask, response)
+         } catch {
+             do {
+                 /// If retry fails (retryCount is 0 or Task.sleep thrown), catch the error and process it with `ErrorProcessing` plugins.
+                 try await sleepIfRetry(
+                    for: error,
+                    endpointRequest: endpointRequest,
+                    retryConfiguration: retryConfiguration
+                 )
+                 
+                 return try await downloadRequest(
+                    endpointRequest,
+                    resumableData: resumableData,
+                    retryConfiguration: retryConfiguration
+                 )
+             } catch {
+                 /// error processing
+                 throw await errorProcessors.process(error, for: endpointRequest)
+             }
+         }
+     }
+
+    func updateTasks() {
+        Task {
+            for task in await allTasks where downloadStateDict[task] == nil {
+                /// In case there is no DownloadState for a given task in the dictionary, we need to create one.
+                downloadStateDict[task] = .init(task: task)
+                
+                /// We need to observe URLSessionTask.State via KVO individually for each task, because there is no delegate callback for the state change.
+                taskStateCancellables[task] = task
+                    .publisher(for: \.state)
+                    .sink { [weak self] state in
+                        self?.downloadStateDict[task]?.taskState = state
+                        
+                        if state == .completed {
+                            self?.taskStateCancellables[task] = nil
+                        }
+                    }
+            }
+        }
+    }
+}
+
+// MARK: URLSession Delegate
+extension DownloadAPIManager: URLSessionDelegate, URLSessionDownloadDelegate {
+    public func urlSession(_: URLSession, downloadTask: URLSessionDownloadTask, didWriteData _: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
+        downloadStateDict[downloadTask]?.totalBytesWritten = totalBytesWritten
+        downloadStateDict[downloadTask]?.totalBytesExpectedToWrite = totalBytesExpectedToWrite
+    }
+
+    public func urlSession(_: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
+        downloadStateDict[downloadTask]?.downloadedFileURL = location
+        updateTasks()
+    }
+
+    public func urlSession(_: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        downloadStateDict[task]?.error = error
+        updateTasks()
+    }
+}
+
+// MARK: URLSessionTask + asyncResponse
+private extension URLSessionTask {
```

Comment:
why we haven't put this wrapper into URLSessionTask extension we have and we keep it private here?

Reply:
agreed, moved and kept internal